### ISSUE TITLE
Add a config that uses the html tokenizer

### ISSUE THREAD

**Issue by NixBiks** [pmbaumgartner/spacy-html-tokenizer#2]
It would be great with a config example that uses the html tokenizer

**Comment by pmbaumgartner** [comment]
Great idea! I did use the entry point for tokenizers, but didn't document it. This means you should be able to do the following in a config:


```
[nlp.tokenizer]
@tokenizers = "html_tokenizer"
```

**Comment by NixBiks** [comment]
Forgot to follow up on this but you actually can't do that. You'll get

```
     54 def create_tokenizer(nlp) -> Tokenizer:
     55     tokenizer = HTMLTokenizer(
     56         vocab=nlp.vocab,
---> 57         rules=nlp.tokenizer.rules,
     58         prefix_search=nlp.tokenizer.prefix_search,
     59         suffix_search=nlp.tokenizer.suffix_search,
     60         infix_finditer=nlp.tokenizer.infix_finditer,
     61         token_match=nlp.tokenizer.token_match,
     62         url_match=nlp.tokenizer.url_match,
     63         unwrap_tags=unwrap_tags,
     64         remove_tags=remove_tags,
     65     )
     66     return tokenizer

AttributeError: 'English' object has no attribute 'tokenizer'
```

Instead I have a callback registered and apply it as explained [here](https://spacy.io/usage/training/#custom-tokenizer).

However then I get a problem during training:

>  [E949] Unable to align tokens for the predicted and reference docs. It is only possible to align the docs when both texts are the same except for whitespace and capitalization. The predicted tokens start with: ['Financial', 'Report', 'October', '-', 'December', '2014', 'Strong', 'EPS', 'growth', 'and']. The reference tokens start with: ['<', 'h1', '>', 'Financial', 'Report', 'October', '-', 'December', '2014</h1><body><p><span><strong', '>']

I assume it's because `nlp.make_doc` does not apply callback when I create `train.spacy` and `dev.spacy`. I have a convert script where I create my `nlp` from a config file with a callback. Then I use `nlp.make_doc` to convert (callback has not been run at this point). Now I start training which makes the callback fire and I get the above error. I'm not sure how the workflow is meant to be in a usecase like this. Should I manually apply the callback before converting?

**Comment by NixBiks** [comment]
Another follow up. Now I'm manually applying the before_init callback before calling `nlp.make_doc` but then I get another error which looks like a bug to me!? 

> ValueError: [E949] Unable to align tokens for the predicted and reference docs. It is only possible to align the docs when both texts are the same except for whitespace and capitalization. The predicted tokens start with: ['Production', 'update', 'September', '2015', 'Tethys', "'", 'share', 'of', 'the', 'production']. The reference tokens start with: ['Production', 'update', 'September', '2015', 'Tethys', "'", 'share', 'of', 'the', 'production']

Let me know if I should move this to spacy issues. I've attached a snippet of my convert script




```python
def apply_before_init(nlp: Language):
    # replicating https://github.com/explosion/spaCy/blob/357be2614effdfdff858eca0eb5f17e167ceb265/spacy/language.py#L1283
    from spacy.schemas import ConfigSchemaInit
    from spacy.util import registry

    config = nlp.config.interpolate()
    # These are the settings provided in the [initialize] block in the config
    I = registry.resolve(config["initialize"], schema=ConfigSchemaInit)
    before_init = I["before_init"]
    if before_init is not None:
        before_init(nlp)


@app.command()
def convert(
    input_path: Path,
    train_path: Path,
    dev_path: Path,
    test_path: Path,
    config_path: Path,
):
    annotations = [Annotation.parse_obj(d) for d in srsly.read_jsonl(input_path)]
    typer.echo(f"Converting {len(annotations):,d} annotations")
    nlp = load_model_from_config(load_config(config_path))
    apply_before_init(nlp)
    train, dev, test = DocBin(), DocBin(), DocBin()
    for annotation in annotations:
        doc = nlp.make_doc(annotation.html)
        doc.cats = annotation.cats
        if annotation.type == "train":
            train.add(doc)
        elif annotation.type == "dev":
            dev.add(doc)
        elif annotation.type == "test":
            test.add(doc)

    for doc_bin, output_path in zip(
        [train, dev, test], [train_path, dev_path, test_path]
    ):
        typer.echo(f"Saving {len(doc_bin):,d} documents to {output_path}")
        doc_bin.to_disk(output_path)
```

**Comment by pmbaumgartner** [comment]
I don't think the first thing is a spaCy issue, it's more likely an issue with this tokenizer. I should have some time to look into this next week.

**Comment by NixBiks** [comment]
Cool - I just noticed the following fails (each .make_doc seems to remove a single `amp;`. Odd)

```python
html = "Mitsui E&amp;amp;P Middle East"
assert nlp.make_doc(html) == nlp.make_doc(nlp.make_doc(html).text)
```

**Comment by NixBiks** [comment]
But I also think it will be necessary not to use callback to set the tokenizer since it won't be called when loading the model (only before training AFAIK)

**Comment by pmbaumgartner** [comment]
I think you found the main issue - `make_doc` should return the same tokenized doc, but it obviously does not because it's calling `clean` again on already "cleaned" text, and modifying it again.

I'm sort of abusing the tokenizer here, as it really shouldn't be handling this stuff, but I'll try and think of a solution for this.

**Comment by NixBiks** [comment]
Yeah it's kinda of a hybrid between preprocessing and tokenization

**Comment by pmbaumgartner** [comment]
> Cool - I just noticed the following fails (each .make_doc seems to remove a single `amp;`. Odd)
> 
> ```python
> html = "Mitsui E&amp;amp;P Middle East"
> assert nlp.make_doc(html) == nlp.make_doc(nlp.make_doc(html).text)
> ```

So I don't think 2 docs will ever be equal. But if you take the text from both of these they should be equal.

In your case what's happening is that `&amp;amp;` is causing an issue in the 'Mitsui E&amp;amp;P Middle East' string. Namely, the first `&amp;` is getting correctly parsed to a `&`, but then there's now a leftover `&amp;` that will then get parsed **again** when it's passed through the tokenizer for training, so there are now two different strings.

```
In [45]: html
Out[45]: 'Mitsui E&amp;amp;P Middle East'

In [46]: nlp(html).text
Out[46]: 'Mitsui E&amp;P Middle East'

In [47]: nlp(nlp(html).text)
Out[47]: Mitsui E&P Middle East
```

So, the main problem is that this isn't really correct HTML, but I think it's also a lot to assume that in order for things to work correctly users will pass correct html without this issue. So I think I need a way to somehow say "this doc has already been processed, don't reprocess it" with the tokenizer, but that would require saving some state in what should be a stateless component. Hmm... 🤔

**Comment by NixBiks** [comment]
Ah yes meant to have 

```python
assert nlp.make_doc(html).text == nlp.make_doc(nlp.make_doc(html).text).text
```

Hm yeah. Quite the change to make it stateful - but of course you do have two different states with this tokenizer (since the text is changed when tokenized which is different from the standard tokenizers)

Think think 🤔
